pkg/dockerregistry/server/pullthroughblobstore.go

```diff
@@ -101,11 +91,35 @@ func (pbs *pullthroughBlobStore) Get(ctx context.Context, dgst digest.Digest) ([
 	return pbs.repo.remoteBlobGetter.Get(ctx, dgst)
 }
 
-// setResponseHeaders sets the appropriate content serving headers
-func setResponseHeaders(w http.ResponseWriter, length int64, mediaType string, digest digest.Digest) {
-	w.Header().Set("Content-Type", mediaType)
-	w.Header().Set("Docker-Content-Digest", digest.String())
-	w.Header().Set("Etag", digest.String())
+// serveRemoteContent tries to use http.ServeContent for remote content.
+func serveRemoteContent(rw http.ResponseWriter, req *http.Request, desc distribution.Descriptor, remoteReader io.ReadSeeker) (bool, error) {
+	// Set the appropriate content serving headers.
+	rw.Header().Set("Content-Type", desc.MediaType)
+	rw.Header().Set("Docker-Content-Digest", desc.Digest.String())
+	rw.Header().Set("Etag", desc.Digest.String())
+
+	// Fallback to Copy if request wasn't given.
+	if req == nil {
+		return false, nil
+	}
+
+	// Check whether remoteReader is seekable. The remoteReader' Seek method must work: ServeContent uses
+	// a seek to the end of the content to determine its size.
+	if _, err := remoteReader.Seek(0, os.SEEK_END); err != nil {
```

Comment:
@smarterclayton This test is about the acknowledgment that the `remoteReader` is really `io.ReadSeeker`. [Acording documentation](https://golang.org/pkg/net/http/#ServeContent) the `http.ServeContent` uses a seek to the end of the content to determine its size.

Basically, the problem hidden deeply inside the docker's client library. They [return](https://github.com/docker/distribution/blob/master/registry/client/transport/http_reader.go#L32) an object with `io.ReadSeeker` interface which in fact [isn't seekable](https://github.com/docker/distribution/blob/master/registry/client/transport/http_reader.go#L120). I can understand them, because if they [don't know the final size of the content](https://github.com/docker/distribution/blob/master/registry/client/transport/http_reader.go#L198-L240), it's impossible to move through it.

So the regression can't happen until the `http.ServeContent` accepts `io.ReadSeeker` as argument. If it stops doing that, then it will regress in golang.

/cc @mfojtik @dmage